File: packages/typescript/src/client/PokeForgeClient.ts
import { HttpClient } from "./http/HttpClient";
import type { PokeForgeClientConfig } from "./http/types";
import { CardsResource } from "./resources/CardsResource";
import { SetsResource } from "./resources/SetsResource";
import { SeriesResource } from "./resources/SeriesResource";
import { CollectionsResource } from "./resources/CollectionsResource";
import { FavoritesResource } from "./resources/FavoritesResource";

/**
 * The main PokeForge API client.
 *
 * @example
 * ```ts
 * import { PokeForgeClient } from '@pokeforge/sdk';
 *
 * const client = new PokeForgeClient({
 *   baseUrl: 'https://api.pokeforge.gg',
 *   auth: { type: 'static', token: 'your-jwt-token' }
 * });
 *
 * // List cards
 * const cards = await client.cards.list({ rarity: 'Rare' });
 *
 * // Get a single card
 * const card = await client.cards.get('card-uuid');
 *
 * // Iterate through all cards
 * for await (const card of client.cards.list()) {
 *   console.log(card.name);
 * }
 * ```
 */
export class PokeForgeClient {
  private readonly http: HttpClient;

  /** Cards API - browse and search Pokemon cards */
  readonly cards: CardsResource;

  /** Sets API - browse card sets */
  readonly sets: SetsResource;

  /** Series API - browse card series */
  readonly series: SeriesResource;

  /** Collections API - manage user card collections */
  readonly collections: CollectionsResource;

  /** Favorites API - manage favorite cards */
  readonly favorites: FavoritesResource;

  /**
   * Create a new PokeForge client.
   *
   * @param config - Client configuration
   */
  constructor(config: PokeForgeClientConfig) {
    this.http = new HttpClient(config);

    // Initialize all resource classes
    this.cards = new CardsResource(this.http);
    this.sets = new SetsResource(this.http);
    this.series = new SeriesResource(this.http);
    this.collections = new CollectionsResource(this.http);
    this.favorites = new FavoritesResource(this.http);
  }

  /**
   * Create a new client instance with updated configuration.
   * Useful for changing authentication mid-session.
   *
   * @param config - Partial configuration to merge
   * @returns A new client instance
   */
  withConfig(config: Partial<PokeForgeClientConfig>): PokeForgeClient {
    const currentConfig = this.http.getConfig();
    return new PokeForgeClient({ ...currentConfig, ...config });
  }

  /**
   * Health check endpoint.
   * Throws if the API is not reachable.
   */
  async health(): Promise<void> {
    await this.http.get("/health");
  }
}

File: packages/csharp/src/PokeForge.SDK/Client/PokeForgeClient.cs
namespace PokeForge.SDK.Client;

using PokeForge.SDK.Client.Http;
using PokeForge.SDK.Client.Resources;

/// <summary>
/// The main PokeForge API client.
/// </summary>
/// <example>
/// <code>
/// using var client = new PokeForgeClient(new PokeForgeClientOptions
/// {
///     AuthToken = "your-jwt-token"
/// });
///
/// // List cards with pagination
/// var page = await client.Cards.ListAsync(new ListCardsOptions { Rarity = "Rare" });
///
/// // Iterate through all cards
/// await foreach (var card in client.Cards.ListAllAsync(new ListCardsOptions { Search = "Pikachu" }))
/// {
///     Console.WriteLine(card.Name);
/// }
/// </code>
/// </example>
public sealed class PokeForgeClient : IDisposable
{
    private readonly HttpClientWrapper _http;
    private bool _disposed;

    /// <summary>
    /// Cards API - browse and search Pokemon cards.
    /// </summary>
    public CardsResource Cards { get; }

    /// <summary>
    /// Sets API - browse card sets.
    /// </summary>
    public SetsResource Sets { get; }

    /// <summary>
    /// Series API - browse card series.
    /// </summary>
    public SeriesResource Series { get; }

    /// <summary>
    /// Collections API - manage user card collections.
    /// Requires authentication.
    /// </summary>
    public CollectionsResource Collections { get; }

    /// <summary>
    /// Favorites API - manage favorite cards.
    /// Requires authentication.
    /// </summary>
    public FavoritesResource Favorites { get; }

    /// <summary>
    /// Artists API - browse card artists.
    /// </summary>
    public ArtistsResource Artists { get; }

    /// <summary>
    /// Blog API - read blog posts.
    /// </summary>
    public BlogResource Blog { get; }

    /// <summary>
    /// Creates a new PokeForge client with default options.
    /// </summary>
    public PokeForgeClient() : this(new PokeForgeClientOptions())
    {
    }

    /// <summary>
    /// Creates a new PokeForge client with the specified options.
    /// </summary>
    /// <param name="options">Client configuration options.</param>
    public PokeForgeClient(PokeForgeClientOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);

        _http = new HttpClientWrapper(options);

        Cards = new CardsResource(_http);
        Sets = new SetsResource(_http);
        Series = new SeriesResource(_http);
        Collections = new CollectionsResource(_http);
        Favorites = new FavoritesResource(_http);
        Artists = new ArtistsResource(_http);
        Blog = new BlogResource(_http);
    }

    /// <summary>
    /// Creates a new client with updated configuration.
    /// The original client is not affected.
    /// </summary>
    /// <param name="configure">Action to configure the new options.</param>
    /// <returns>A new client with the updated configuration.</returns>
    public PokeForgeClient WithOptions(Action<PokeForgeClientOptions> configure)
    {
        var newOptions = _http.Options.Clone();
        configure(newOptions);
        return new PokeForgeClient(newOptions);
    }

    /// <summary>
    /// Performs a health check against the API.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task HealthAsync(CancellationToken cancellationToken = default)
    {
        await _http.GetAsync<object>("/health", cancellationToken: cancellationToken)
            .ConfigureAwait(false);
    }

    /// <summary>
    /// Disposes the client and releases resources.
    /// If a custom HttpClient was provided, it is NOT disposed.
    /// </summary>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _http.Dispose();
    }
}
